src/ffi.rs

```diff
@@ -578,6 +578,199 @@ pub unsafe extern "C" fn context_get_result(
+///
+/// - `atc` must be a valid pointer to a C-style string, must be properly aligned,
+///    and must not have '\0' in the middle.
+/// - `schema` must be a valid pointer returned by [`schema_new`].
+/// - `fields_buf` must be a valid to write for `fields_len * size_of::<u8>()` bytes,
+///    and it must be properly aligned if non-null.
+/// - `fields_len` must be a valid to write for `size_of::<usize>()` bytes,
+///    and it must be properly aligned if non-null.
+/// - `fields_total` must be a valid to write for `size_of::<usize>()` bytes,
+///    and it must be properly aligned if non-null.
+/// - `operators` must be a valid to write for `size_of::<u64>()` bytes,
+///    and it must be properly aligned if non-null.
+/// - `errbuf` must be valid to read and write for `errbuf_len * size_of::<u8>()` bytes,
+///    and it must be properly aligned.
+/// - `errbuf_len` must be valid to read and write for `size_of::<usize>()` bytes,
+///    and it must be properly aligned.
+/// - If `fields_buf` is non-null, `fields_len` and `fields_total` must be non-null.
+/// - If `fields_buf` is null, `fields_len` and `fields_total` can be non-null
+///   for writing required buffer length and total number of fields.
+#[cfg(feature = "expr_validation")]
+#[no_mangle]
+pub unsafe extern "C" fn expression_validate(
+    atc: *const u8,
+    schema: &Schema,
+    fields_buf: *mut u8,
+    fields_len: *mut usize,
+    fields_total: *mut usize,
+    operators: *mut u64,
+    errbuf: *mut u8,
+    errbuf_len: *mut usize,
+) -> i64 {
+    use std::collections::HashMap;
+
+    use crate::ast::{BinaryOperatorFlags, Expression, LogicalExpression};
+    use crate::parser::parse;
+    use crate::semantics::FieldCounter;
+    use crate::semantics::Validate;
+
+    let atc = ffi::CStr::from_ptr(atc as *const c_char).to_str().unwrap();
+    let errbuf = from_raw_parts_mut(errbuf, ERR_BUF_MAX_LEN);
+
+    // Parse the expression
+    let result = parse(atc).map_err(|e| e.to_string());
+    if let Err(e) = result {
+        let errlen = min(e.len(), *errbuf_len);
+        errbuf[..errlen].copy_from_slice(&e.as_bytes()[..errlen]);
+        *errbuf_len = errlen;
+        return ATC_ROUTER_EXPRESSION_VALIDATE_FAILED;
+    }
+    // Unwrap is safe since we've already checked for error
+    let ast = result.unwrap();
+
+    // Validate expression with schema
+    if let Err(e) = ast.validate(schema).map_err(|e| e.to_string()) {
+        let errlen = min(e.len(), *errbuf_len);
+        errbuf[..errlen].copy_from_slice(&e.as_bytes()[..errlen]);
+        *errbuf_len = errlen;
+        return ATC_ROUTER_EXPRESSION_VALIDATE_FAILED;
+    }
+
+    // Get used fields
+    if !(fields_buf.is_null() && fields_len.is_null() && fields_total.is_null()) {
+        if !fields_buf.is_null() {
+            assert!(
+                !fields_len.is_null() && !fields_total.is_null(),
+                "fields_len and fields_total must be non-null when fields_buf is non-null"
+            );
+        }
```

Comment:
```suggestion
    if !fields_buf.is_null() {
        assert!(
          !fields_len.is_null() && ! fields_total.is_null(),
          "fields_len and fields_total must be non-null when fields_buf is non-null"
        )
```

Reply:
This would not allow a null `fields_buf` with non-null `fields_len` or `fields_total` to get the required buffer length and total field count without a valid buffer. This behavior is indeed NOT encouraged (since a waste of validating and getting fields/operators is called) but I think it's necessary to provide that form of interface to make it more flexible and keep consistency to other interfaces like `get_fields`.